### PR TITLE
ci: cache all build tasks

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-2710

Improvement for our 20% tech debt initiative.

This PR should allow us to confidently use `nx run` to run any command in the repository. There's still some start-up time that may make it tempting to use `npm run`, but it shouldn't be as bad as before.

In this PR, I made the following changes:
1. Added caching to all `build` tasks in the repository.
   * By default, new packages won't be cached.
2. Added default caching behaviour that should work for most cases.
3. Replaced caching behaviour in CI (this ensures we don't break caching).
   * We now cache Nx's `.cache/nx` directory instead of the output of every package.
     * Jobs now start without the build of every project. Using Nx, they load or build every package they need.
   * We now cache the entire `node_modules` directory instead of NPM's cache.
     * I can revert this change, but I noticed that it results in a huge reduction in CI run time.
   * Releases are still built from scratch.
4. Added docs on how to cache builds.
5. Made `dev` scripts automatically build dependencies.

TODO:
- [x] Revert `ui-kit.code-workspaces` change.
- [x] Fix bundle size report performance

### CODE REVIEW

File: .github/actions/build/action.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+name: 'Install, build & cache'
```
Comment (btaillon) [inline .github/actions/build/action.yml:1]: The different between this and `setup`, is that this is meant to create the cache, while the setup task now requires the cache.

File: .github/actions/setup/action.yml [hunk 2]
```diff
@@ -1,51 +1,43 @@
...
+      shell: bash
     - name: Install npm
       run: npm i -g npm@9
       shell: bash
     - uses: actions/cache@v3
-      id: build-cache
+      id: npm-cache
       with:
         path: |
-          packages/bueno/dist
```
Comment (btaillon) [inline .github/actions/setup/action.yml:16]: Goodbye to this list

File: .github/actions/setup/action.yml [hunk 3]
```diff
@@ -1,51 +1,43 @@
...
-          packages/atomic-hosted-page/dist
-          packages/atomic-angular/projects/atomic-angular/dist
-          packages/samples/headless-react
-          packages/samples/iife/www/cdn
-          packages/samples/headless-ssr/.next
-          packages/quantic/force-app/main/default/staticresources/coveoheadless
-          packages/quantic/force-app/main/default/staticresources/coveobueno
-          utils/atomic-storybook/storybook-static
-          !node_modules
-          !**/node_modules
```
Comment (btaillon) [inline .github/actions/setup/action.yml:37]: Say goodbye to this list.

File: nx.json [hunk 4]
```diff
@@ -55,5 +108,6 @@
       "dependsOn": ["release:build", "release:phase1"]
     }
   },
-  "defaultBase": "master"
+  "defaultBase": "master",
+  "$schema": "./node_modules/nx/schemas/nx-schema.json"
```
Comment (btaillon) [inline nx.json:112]: added by `nx repair`.

File: package.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
     "postinstall": "husky install && patch-package && nx run-many --target=install:isolated",
     "reset:install": "git checkout origin/master package-lock.json && npm ci",
     "clean:install": "rimraf **/node_modules && npm run reset:install",
-    "dev:atomic": "concurrently \"nx run headless:dev\" \"nx run atomic:dev\"",
+    "dev:atomic": "concurrently \"npm run -w @coveo/headless dev\" \"npm run -w @coveo/atomic dev\"",
```
Comment (btaillon) [inline package.json:8]: Without this change, this would run Nx twice in parallel, which could cause conflicts.

Nx has some mechanisms to handle dev tasks, but they are more complex. IMO, it wouldn't be a valuable use of our time.

File: packages/auth/project.json [hunk 6]
```diff
@@ -1 +1,18 @@
-{}
+{
+  "name": "auth",
+  "$schema": "../../node_modules/nx/schemas/project-schema.json",
+  "targets": {
+    "cached:build": {
+      "executor": "nx:run-commands",
+      "options": {
+        "commands": ["npm run build:bundles", "npm run build:definitions"],
+        "parallel": true,
```
Comment (btaillon) [inline packages/auth/project.json:9]: I took some easy-wins parallelizing like this.

File: ui-kit.code-workspace [hunk 7]
```diff
@@ -108,7 +108,6 @@
     "jest.autoRevealOutput": "off",
     "npm.packageManager": "npm",
     "files.exclude": {
-      "**/node_modules/": true,
```
Comment (btaillon) [inline ui-kit.code-workspace:111]: will revert

**Review by olamothe — APPROVED** [review]
Lots of, but most of it seems straightforward enough.

Plus, if everything builds on CI, then there's not much else to say !
💯

File: .github/actions/e2e-atomic-angular/action.yml [hunk 8]
```diff
@@ -16,9 +16,8 @@ runs:
       name: Run Cypress
       with:
         browser: chrome
-        command-prefix: 'npx cypress-repeat run -n 3 --until-passes --rerun-failed-only --'
-        working-directory: ./packages/samples/angular
-        start: npm run dev
+        command-prefix: 'npx -w @coveo/atomic-angular-samples cypress-repeat run -n 3 --until-passes --rerun-failed-only --'
```
Comment (mrrajamanickam-coveo) [inline .github/actions/e2e-atomic-angular/action.yml:19]: Would it be possible to extract common commands such as `cypress-repeat run -n 3 --until-passes --rerun-failed-only` to common nx config?

File: package.json (hunk 5 above)
Comment (mrrajamanickam-coveo) [reply package.json:8]: Can't `nx run-many --target=dev` be used with selected workspaces/projects?

File: packages/samples/atomic-react/project.json [hunk 9]
```diff
@@ -1,3 +1,22 @@
...
+      "outputs": ["{projectRoot}/public/dist"],
+      "executor": "nx:run-commands",
+      "options": {
+        "commands": ["tsc --noEmit", "tsc --module commonjs --noEmit"],
+        "parallel": true,
+        "cwd": "packages/samples/atomic-react"
+      }
+    },
+    "build": {
+      "dependsOn": ["cached:build"],
```
Comment (mrrajamanickam-coveo) [inline packages/samples/atomic-react/project.json:18]: Can we mark atomic build as a dep here ?

File: packages/samples/headless-react/project.json [hunk 10]
```diff
@@ -1,3 +1,18 @@
...
+    "cached:build": {
+      "executor": "nx:run-commands",
+      "options": {
+        "commands": ["npm run build:client", "npm run build:server"],
+        "parallel": true,
+        "cwd": "packages/samples/headless-react"
+      }
+    },
+    "build": {
+      "dependsOn": ["cached:build"],
```
Comment (mrrajamanickam-coveo) [inline packages/samples/headless-react/project.json:14]: Can we mark headless build as a dependency for all samples/headless-* pkgs?

File: packages/samples/headless-ssr/next.config.js [hunk 11]
```diff
@@ -1,4 +1,4 @@
 /** @type {import('next').NextConfig} */
-const nextConfig = {}
+const nextConfig = {experimental: {appDir: true}};
```
Comment (mrrajamanickam-coveo) [inline packages/samples/headless-ssr/next.config.js:2]: Why is this required, esp with the changes in this PR? Doesn't nextjs 13 default to app dir?

**Comment by mrrajamanickam-coveo** [comment]
For this PR
>Total duration
[53m 36s](https://github.com/coveo/ui-kit/actions/runs/5980457404/usage)

vs PRs in master are around ~1h 10m
>[1h 8m 19s](https://github.com/coveo/ui-kit/actions/runs/5968020857/usage) 

Am I looking at this correctly?

File: .github/actions/e2e-atomic-angular/action.yml (hunk 8 above)
Comment (btaillon) [reply .github/actions/e2e-atomic-angular/action.yml:19]: Yes and no.

Technically, Nx doesn't allow us to create "default" tasks. However, we have an Nx plugin in our workspace (under `utils/nx-plugin`) which already generates Nx tasks, so it wouldn't be much more work to add a default `cypress-repeat` task.

I'll give it a shot.

File: packages/samples/atomic-react/project.json (hunk 9 above)
Comment (btaillon) [reply packages/samples/atomic-react/project.json:18]: There's no need. This task depends on the `atomic-react:cached:build` task, which itself depends on the build task of all dependencies (`^build`).

File: packages/samples/headless-react/project.json (hunk 10 above)
Comment (btaillon) [reply packages/samples/headless-react/project.json:14]: Same as above, there's no need since `headless-react-samples:cached:build` already depends on `^build` (by default, inherited from `nx.json`).

File: packages/samples/headless-ssr/next.config.js (hunk 11 above)
Comment (btaillon) [reply packages/samples/headless-ssr/next.config.js:2]: Had an error with Next.js, don't think this bit is necessary anymore, forgot to revert it 😅

**Comment by btaillon** [comment]
> For this PR
> 
> > Total duration
> > [53m 36s](https://github.com/coveo/ui-kit/actions/runs/5980457404/usage)
> 
> vs PRs in master are around ~1h 10m
> 
> > [1h 8m 19s](https://github.com/coveo/ui-kit/actions/runs/5968020857/usage)
> 
> Am I looking at this correctly?

Yes, but most of the duration in both cases seems to be due to the first Atomic e2e test taking like 45+ mins to run.
